### PR TITLE
feat: made BaseNode.parent non-optional

### LINKED ISSUE #1417: Reworking the `parent` property on AST Nodes
Right now the `parent` property is optional on all AST nodes.
However in reality, it's always defined on all AST nodes, except for `Program`, where it is missing.

This leads to a bunch of unnecessary ifs/optional chains/"trust me it's right" assertions, all of which "harm" the type safety, or are impossible to get coverage on (leading to istanbul ignore statements):

```ts
// this shouldn't happen, checking just in case
/* istanbul ignore if */ if (node.parent) {
  // ...
}

// selector asserts this is correct
const parent = node.parent as TSESTree.VariableDeclaration;

if (node.parent?.type === AST_NODE_TYPES.VariableDeclaration) {
  // ...
}
```

There is an issue with this, however.
`typescript-estree` does not assign the parent property; this is done by ESLint as it traverses the AST at lint time.

So there are two solutions I can see:
1) change `typescript-estree` so that it sets the `parent` appropriately.
2) create a separate set of types for use within plugins, which correctly type the `parent` property.

---

As an aside, doing this would unlock a really awesome benefit wherein we could make all the types generic, allowing you to specify the expected parent type ahead of time:
```ts
interface BaseNode<TParent extends Node = Node> {
  // ...
  parent: TParent;
}

interface VariableDeclarator<TParent extends Node = Node> extends BaseNode<TParent> {}

interface Program extends BaseNode {
  parent?: never;
}
```

This would be great for things like return values of functions which explicitly check parent types, and for eslint selectors that explicitly state the tree structure (`VariableDeclaration > VariableDeclarator`).

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
BREAKING CHANGE:
changes the AST types in a non-backwards compatible way.

---

## PR Checklist

-   [x] Addresses an existing open issue: fixes #1417
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Doesn't quite make `BaseNode`'s `parent` property generic (a good next step?), but does make `Program`'s property `parent?: never`.

I removed all checks like `/if \(\w+\.parent/`, `/!\w+\.parent/`, and `/\w+\.parent ((&&)|(\|\|))/`, then added back any that caused type checking complaints. Those were mostly from places that recursively went up the AST via `node.parent`.

### CODE REVIEW

File: packages/ast-spec/src/base/BaseNode.ts [hunk 1]
```diff
@@ -3,12 +3,8 @@ import type { AST_NODE_TYPES } from '../ast-node-types';
...
 
 export interface BaseNode extends NodeOrTokenData {
-  /**
-   * The parent node of the current node
-   *
-   * This is added in the @typescript-eslint/types package as ESLint adds it
-   * while traversing.
-   */
+  // The parent node of the current node
+  // (added in @typescript-eslint/types as ESLint adds it while traversing)
```
Comment (JoshuaKGoldberg) [inline packages/ast-spec/src/base/BaseNode.ts:7]: Switched to an inline `//` comment so the `/** */` jsdoc comment doesn't risk getting attached to `type`.

File: packages/typescript-estree/src/node-utils.ts [hunk 2]
```diff
@@ -223,7 +223,10 @@ export function canContainDirective(
  * @param ast the AST object
  * @returns the range data
  */
-export function getRange(node: ts.Node, ast: ts.SourceFile): [number, number] {
+export function getRange(
+  node: Pick<ts.Node, 'getEnd' | 'getStart'>,
+  ast: ts.SourceFile,
+): [number, number] {
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/src/node-utils.ts:229]: `Program`-typed values were getting passed in here. Having the `node` `Pick` out the only methods it needs stopped some type errors.

File: packages/typescript-estree/src/convert.ts [hunk 3]
```diff
@@ -390,7 +387,7 @@ export class Converter {
       params: typeParameters.map(typeParameter =>
         this.convertType(typeParameter),
       ),
-    };
+    } as TSESTree.TSTypeParameterDeclaration;
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/src/convert.ts:390]: Is there a reason this and the other node don't use `this.createNode`?

**Review by bradzacher — APPROVED** [review]
keanubeautiful

**Pull request merged**